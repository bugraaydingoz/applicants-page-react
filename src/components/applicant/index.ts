export { Applicant } from './Applicant'
export { ApplicantList } from './ApplicantList'
