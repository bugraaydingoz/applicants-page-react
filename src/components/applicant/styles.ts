import styled from 'styled-components'

export const ApplicantContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  margin-right: var(--space-md);
  padding: var(--space-sm) 0;
  width: 305px;
  min-width: 305px;
  height: 241px;
  border: 1px solid var(--gray-200);
  border-radius: var(--border-radius-lg);

  @media (min-width: 768px) {
    overflow: none;
    flex-wrap: wrap;
    margin-bottom: var(--space-md);

    // Remove last row's margin
    &:nth-last-child(1),
    &:nth-last-child(2),
    &:nth-last-child(3),
    &:nth-last-child(4) {
      margin-bottom: 0;
    }
  }
`

export const Applicants = styled.div`
  display: flex;
  flex-direction: row;
  margin-bottom: var(--space-lg);
  margin-right: calc(-1 * var(--space-sm));
  overflow: scroll;

  @media (min-width: 768px) {
    flex-wrap: wrap;
    margin-right: calc(-1 * var(--space-md));
  }

  &:last-child {
    margin-bottom: 0;
  }
`

interface Props {
  color: string
}

export const AvatarContainer = styled.div<Props>`
  display: flex;
  align-items: center;
  justify-content: center;
  margin-bottom: var(--space-xs);
  width: 70px;
  height: 70px;
  border-radius: 999px;
  background-color: ${({ color }) => color};
`

function CSSColor(color: string): string {
  switch (color) {
    case 'yellow':
      return 'var(--yellow-400)'
    case 'gray':
    default:
      return 'var(--gray-400)'
  }
}

export const TagContainer = styled.div<Props>`
  display: flex;
  margin-bottom: var(--space-xs);
  padding: 4px 8px;
  height: 18px;
  font-weight: bold;
  font-size: 12px;
  line-height: 10px;
  text-transform: uppercase;
  color: white;
  background-color: ${({ color }) => CSSColor(color)};
  border-radius: var(--border-radius-md);

  &:last-child {
    margin-bottom: 0;
  }
`
