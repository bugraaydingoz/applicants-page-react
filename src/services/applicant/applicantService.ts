import { Applicant } from './types'
import { applicants } from './data'
import { delayFor, random } from '../utils'

export class ApplicantService {
  static async getAll(): Promise<Applicant[]> {
    // Mock server response time
    await delayFor(random(200, 1200))

    // Mock server error
    if (random(1, 10) < 3) {
      throw new Error("Applicants couldn't fetched")
    }

    return applicants
  }
}
