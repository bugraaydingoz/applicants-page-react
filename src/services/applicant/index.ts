export type { Applicant } from './types'
export { ApplicantStatus } from './types'
export { ApplicantService } from './applicantService'
