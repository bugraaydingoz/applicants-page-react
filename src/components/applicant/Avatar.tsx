import React from 'react'

// Styles
import { AvatarContainer } from './styles'
import { H5 } from '../shared/typography'

const colors = [
  // Purple
  {
    light: '#E3E7F9',
    dark: '#405493',
  },

  // Blue
  {
    light: '#CEF0F4',
    dark: '#20ADC5',
  },

  // Yellow-ish
  {
    light: '#F0F4E1',
    dark: '#ACAF36',
  },

  // Orange
  {
    light: '#FEE8D3',
    dark: '#E97428',
  },

  // Red
  {
    light: '#FCD6DE',
    dark: '#E52A50',
  },
]

interface Props {
  firstName: string
  lastName: string
}

export function Avatar({ firstName, lastName }: Props) {
  const random = () => Math.floor(Math.random() * colors.length - 1) + 1
  const color = colors[random()]
  const initials = `${firstName[0]}${lastName[0]}`
  return (
    <AvatarContainer color={color.light}>
      <H5 color={color.dark}>{initials}</H5>
    </AvatarContainer>
  )
}
