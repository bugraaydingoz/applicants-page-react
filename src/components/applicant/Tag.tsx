import React from 'react'

// Styles
import { TagContainer } from './styles'

interface Props {
  color: 'gray' | 'yellow'
  children: React.ReactNode
}

export function Tag({ color, children }: Props) {
  return <TagContainer color={color}>{children}</TagContainer>
}
