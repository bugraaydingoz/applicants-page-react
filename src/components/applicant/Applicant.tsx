import React from 'react'
import type { Applicant as Props } from '../../services/applicant'

// Components
import { Avatar } from './Avatar'

// Styles
import { ApplicantContainer } from './styles'
import { H5, Body1 } from '../shared/typography'
import { Tag } from './Tag'

export function Applicant(applicant: Props) {
  return (
    <ApplicantContainer>
      <Avatar firstName={applicant.firstName} lastName={applicant.lastName} />
      <H5>{`${applicant.firstName} ${applicant.lastName}`}</H5>
      <Body1>{applicant.phoneNumber}</Body1>
      <Body1 style={{ marginBottom: 'var(--space-xs)' }}>
        {applicant.emailAddress}
      </Body1>

      {applicant.appointmentISO && (
        <Tag color="gray">{applicant.appointmentISO}</Tag>
      )}

      {/* {applicant.viewedISO && <Tag color="gray">{applicant.viewedISO}</Tag>} */}

      {applicant.bid && <Tag color="yellow">{applicant.bid}</Tag>}
    </ApplicantContainer>
  )
}
