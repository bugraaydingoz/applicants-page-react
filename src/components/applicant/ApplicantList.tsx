import React from 'react'
import type { Applicant as ApplicantType } from '../../services/applicant'

// Components
import { Applicant } from './Applicant'

// Styles
import { Applicants, ApplicantMock, LoadingBar } from './styles'
import { H4 } from '../shared/typography'

interface Props {
  title: string
  applicants: ApplicantType[]
  loading: boolean
}

export function ApplicantList({ title, applicants, loading }: Props) {
  if (loading) {
    return (
      <>
        <LoadingBar />
        <Applicants>
          <ApplicantMock />
          <ApplicantMock />
          <ApplicantMock />
          <ApplicantMock />
        </Applicants>
      </>
    )
  }

  if (!applicants.length) {
    return <></>
  }

  return (
    <>
      <H4
        style={{ marginBottom: 'var(--space-md)' }}
      >{`${title} (${applicants.length})`}</H4>

      <Applicants>
        {applicants.map((applicant) => (
          <Applicant key={applicant.id} {...applicant} />
        ))}
      </Applicants>
    </>
  )
}
