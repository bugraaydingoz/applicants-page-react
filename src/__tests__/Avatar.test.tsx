import React from 'react'
import renderer from 'react-test-renderer'
import { Avatar } from '../components/applicant/Avatar'

test('renders correctly', () => {
  const tree = renderer
    .create(<Avatar firstName="Bugra" lastName="Aydingoz" />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})
