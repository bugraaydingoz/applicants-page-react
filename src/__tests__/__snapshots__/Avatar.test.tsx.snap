// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sc-AxhUy gWBIVe"
  color="#E3E7F9"
>
  <h5
    className="sc-AxheI iQGnXX"
    color="#405493"
  >
    BA
  </h5>
</div>
`;
